### PR TITLE
Read data from SSL Game Controller

### LINKED ISSUE #72: Read data from SSL-Refbox
In addition to SSL vision data that gives us the state of the world (robot positions, etc.) we also need to listen for commands from the Referee Box.

The refbox from previous years can be found here: https://github.com/RoboCup-SSL/ssl-refbox
There had been talk about re-writing it for RoboCup 2019. The new repository can be found here: https://github.com/RoboCup-SSL/ssl-game-controller

There is sample client code in the RefBox repositories. Our code can likely be based very closely on that, and will likely be very similar to our code that receives SSL-Vision data (see the SSLVisionClient in the `backend_input` node)

Data that is received should be stored in the datastructure created in #43 , so this ticket is partially blocked by #43 . Nearly all the work can be done without the datastructure however.

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description
Read the game state from the protobuf messages emitted by SSL Game Controller (Refbox replacement) and provide it to the AI Game State class.

### Testing Done

Tested with SSL Game Controller. (No unit tests are there for any other network_input classes)

### Resolved Issues

Resolves #72 

### Length Justification

A large chunk of this PR is a protobuf message copypasted from ssl-game-controller

### Review Checklist
*(Please check every item to indicate your code complies with it (by changing `[ ]`->`[x]`). This will hopefully save both you and the reviewer(s) a lot of time!)*
***It is the reviewers responsibility to also make sure every item here has been covered***
- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom` 
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

***Feel free to make additions of things that we should be checking to this file if you think there's something missing!!!!***

### CODE REVIEW

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        // Start listening again to receive the next data
+        socket_.async_receive_from(
+            boost::asio::buffer(raw_received_data_, max_buffer_length), sender_endpoint_,
+            boost::bind(&SSLGameControllerClient::handleDataReception, this,
+                        boost::asio::placeholders::error,
+                        boost::asio::placeholders::bytes_transferred));
+
+        // TODO: Log a proper warning here so we can be notified that there was an error
+        // while handling received data. Can be done once
+        // https://github.com/UBC-Thunderbots/Software/issues/33 is done
```
Comment (garethellis0) [inline src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp]: I believe #33 is resolved now, so we can do this? See `ssl_vision_client.cp` for an example.

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+     *
+     * @param error The error code obtained when receiving the incoming data
+     * @param num_bytes_received How many bytes of data were received
+     */
+    void handleDataReception(const boost::system::error_code &error,
+                             size_t num_bytes_received);
+
+    boost::asio::io_service io_service;
+    boost::asio::ip::udp::socket socket_;
+    boost::asio::ip::udp::endpoint sender_endpoint_;
```
Comment (garethellis0) [inline src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h:50]: Could we add a quick comment for each member function?

File: src/thunderbots/software/proto/src/ssl_referee.proto [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+message ProposedGameEvent {
+    // The UNIX timestamp when the game event proposal will time out, in microseconds.
+    required uint64 valid_until = 1;
+    // The identifier of the proposer.
+    required string proposer_id = 2;
+    // The proposed game event.
+    required GameEvent game_event = 3;
+}
```
Comment (garethellis0) [inline src/thunderbots/software/proto/src/ssl_referee.proto]: nit; github complaining about end-of-file

File: src/thunderbots/software/util/ros_messages.cpp [hunk 4]
```diff
@@ -70,5 +73,13 @@ namespace Util
...
+
+        RefboxGameState createGameStateFromROSMessage(
+            const thunderbots_msgs::RefboxCommand& command)
+        {
+            // the values in RefboxGameState correspond to the constants in
+            // RefboxCommand.msg
+            return (RefboxGameState)command.command;
+        }
     }  // namespace ROSMessages
 }  // namespace Util
```
Comment (garethellis0) [inline src/thunderbots/software/util/ros_messages.cpp]: nit; github complaining about end-of-file

**Review by garethellis0 — CHANGES_REQUESTED** [review]
This is looking good! Nice work.

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    : socket_(io_service)
+{
+    boost::asio::ip::udp::endpoint listen_endpoint(
+        boost::asio::ip::address::from_string(ip_address), port);
+    socket_.open(listen_endpoint.protocol());
+    socket_.bind(listen_endpoint);
+
+    // Join the multicast group.
+    socket_.set_option(boost::asio::ip::multicast::join_group(
+        boost::asio::ip::address::from_string(ip_address)));
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp:24]: Not sure if that's necessary? I'm assuming you want the handling to be logging a proper error message and re-throwing the exception? Giving an outright invalid IP address is a pretty significant failure we'd want to take down the system right away, so I think it's fine to not catch an exception here.

File: src/thunderbots/software/util/ros_messages.cpp [hunk 6]
```diff
@@ -1,5 +1,8 @@
 #include "util/ros_messages.h"
 
+#include "ros_messages.h"
```
Comment (MathewMacDougall) [inline src/thunderbots/software/util/ros_messages.cpp]: Is this including the same thing as the `#include` above? In any case, this should use the absolute path not the relative path.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Just a few small comments but otherwise looks great!

File: src/thunderbots/software/network_input/backend.cpp [hunk 7]
```diff
@@ -131,3 +131,159 @@ std::optional<thunderbots_msgs::Team> Backend::getFilteredEnemyTeamMsg(
...
+    else
+    {
+        refbox_data.us   = yellow;
+        refbox_data.them = blue;
+    }
+
+    return std::make_optional<thunderbots_msgs::RefboxData>(refbox_data);
+}
+
+// this maps a protobuf Referee_Command enum to its ROS message equivalent, for blue team
```
[the remark this answers is not in the document]
Comment (jordanzliu) [reply src/thunderbots/software/network_input/backend.cpp]: done

File: src/thunderbots/software/network_input/backend.h [hunk 8]
```diff
@@ -67,10 +70,25 @@ class Backend
...
+    // global coordinates. To AI, +x will always be enemy and -x will always be friendly.
+    FieldSide our_field_side;
+    // TODO: the rest of backend should be spitting out coordinates transformed as above
+    // https://github.com/UBC-Thunderbots/Software/issues/163
+
+    void setOurFieldSide(bool blue_team_on_positive_half);
+
+    int32_t getTeamCommand(const Referee::Command &command);
+    Point getTeamLocalCoordinates(const Referee::Point &point);
+    thunderbots_msgs::RefboxTeamInfo getTeamInfo(const Referee::TeamInfo &team_info);
```
[the remark this answers is not in the document]
Comment (jordanzliu) [reply src/thunderbots/software/network_input/backend.h:117]: done

File: src/thunderbots/software/network_input/main.cpp [hunk 9]
```diff
@@ -51,6 +55,21 @@ int main(int argc, char** argv)
...
+        ssl_gamecontroller_client = std::make_unique<SSLGameControllerClient>(
+            Util::Constants::SSL_GAMECONTROLLER_MULTICAST_ADDRESS,
+            Util::Constants::SSL_GAMECONTROLLER_MULTICAST_PORT);
+    }
+    catch (const boost::exception& ex)
+    {
+        std::cerr << "An error occured while setting up the SSL Game Controller Client:"
+                  << std::endl
+                  << boost::diagnostic_information(ex) << std::endl;
+        return EXIT_FAILURE;
```
[the remark this answers is not in the document]
Comment (jordanzliu) [reply src/thunderbots/software/network_input/main.cpp:84]: I'll open an issue for this

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp (hunk 1 above)
Comment (jordanzliu) [reply src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp]: added error logging to ssl_gamecontroller_client

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h (hunk 2 above)
Comment (jordanzliu) [reply src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h:50]: i fixed the existing javadoc function comments

File: src/thunderbots/software/proto/src/ssl_referee.proto (hunk 3 above)
Comment (jordanzliu) [reply src/thunderbots/software/proto/src/ssl_referee.proto]: fixed

File: src/thunderbots/software/ai/ai.cpp [hunk 10]
```diff
@@ -1,5 +1,8 @@
 #include "ai.h"
 
+#include <g3log/g3log.hpp>
```
[the remark this answers is not in the document]
Comment (jordanzliu) [reply src/thunderbots/software/ai/ai.cpp]: fixed

File: src/thunderbots/software/util/ros_messages.cpp (hunk 6 above)
Comment (jordanzliu) [reply src/thunderbots/software/util/ros_messages.cpp]: fixed the include

Comment (jordanzliu) [reply src/thunderbots/software/util/ros_messages.cpp]: I think i fixed the thing with the exception too

File: src/thunderbots/software/network_input/main.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/software/network_input/main.cpp:84]: Sounds good! Maybe just leave a quick comment on both instances in this file linking to the new issue.

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp:24]: Fair enough. We'll actually catch and log this exception at the top level anyhow (but then still fail).

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h (hunk 2 above)
Comment (garethellis0) [reply src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h:50]: Sorry! I mis-typed, I meant for each member _variable_.

**Review by MathewMacDougall — COMMENTED** [review]
LGTM pending Gareth's last clarification about member variable comments.

**Comment by MathewMacDougall** [comment]
@jordanzliu looks like some conflicts appeared (not sure how)

**Comment by jordanzliu** [comment]
yeah idk what's going on after i merged in the latest changes from master, but it's not compiling for me now

**Comment by jordanzliu** [comment]
i messed up something with the protobuf messages, for some reason protobuf compile fails silently. 
EDIT: there's naming conflicts between some newly-updated protobuf files from SSL Game Controller
EDIT 2: something is seriously, seriously wrong with protobuf and now it's not linking to the compiled protobuf object files.

**Comment by MathewMacDougall** [comment]
That's super weird. Let me know if you want me to look at anything.

**Comment by garethellis0** [comment]
@jordanzliu - looks like travis kinda choked here. Could you push another commit to force a rebuild?

**Comment by jordanzliu** [comment]
Everything seems to work fine now

**Comment by garethellis0** [comment]
:+1: - looks like there are just a few small comments left, then this should be good to go!

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Sorry, 2 more really small things but otherwise LGTM.

File: src/thunderbots/software/network_input/backend.cpp [hunk 11]
```diff
@@ -146,3 +146,160 @@ std::optional<thunderbots_msgs::Team> Backend::getFilteredEnemyTeamMsg(
...
+    {Referee_Command_TIMEOUT_BLUE, thunderbots_msgs::RefboxCommand::TIMEOUT_THEM},
+    {Referee_Command_TIMEOUT_YELLOW, thunderbots_msgs::RefboxCommand::TIMEOUT_US},
+    {Referee_Command_GOAL_BLUE, thunderbots_msgs::RefboxCommand::GOAL_THEM},
+    {Referee_Command_GOAL_YELLOW, thunderbots_msgs::RefboxCommand::GOAL_US},
+    {Referee_Command_BALL_PLACEMENT_BLUE,
+     thunderbots_msgs::RefboxCommand::BALL_PLACEMENT_THEM},
+    {Referee_Command_BALL_PLACEMENT_YELLOW,
+     thunderbots_msgs::RefboxCommand::BALL_PLACEMENT_US}};
+
+int32_t Backend::getTeamCommand(const Referee::Command &command)
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/backend.cpp:240]: Ah ok. Nevermind then.

File: src/thunderbots/software/network_input/main.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (jordanzliu) [reply src/thunderbots/software/network_input/main.cpp:84]: https://github.com/UBC-Thunderbots/Software/issues/275 
issue opened, added to comments

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h (hunk 2 above)
Comment (jordanzliu) [reply src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h:50]: commented

File: src/thunderbots/software/util/refbox_constants.h [hunk 12]
```diff
@@ -4,24 +4,25 @@
...
+    DIRECT_FREE_THEM     = 9,
+    INDIRECT_FREE_US     = 10,
+    INDIRECT_FREE_THEM   = 11,
+    TIMEOUT_US           = 12,
+    TIMEOUT_THEM         = 13,
+    GOAL_US              = 14,
+    GOAL_THEM            = 15,
+    BALL_PLACEMENT_US    = 16,
+    BALL_PLACEMENT_THEM  = 17,
     LAST_ENUM_ITEM_UNUSED
```
[the remark this answers is not in the document]
Comment (jordanzliu) [reply src/thunderbots/software/util/refbox_constants.h]: renamed

**Review by MathewMacDougall — DISMISSED** [review]
LGTM :+1: Awesome work! :tada:

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada: - nice job!

Looks like we'll need to update this branch though, probably will need to re-approve after.

**Review by MathewMacDougall — APPROVED** [review]

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Pull request merged**